include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+  template <typename VectorType  = PETScWrappers::VectorBase,
+            typename PMatrixType = PETScWrappers::MatrixBase,
+            typename AMatrixType = PMatrixType>
+  class TimeStepper
+  {
+  public:
+    /**
+     * Constructor.
+     */
+    TimeStepper(const TimeStepperData &data     = TimeStepperData(),
+                const MPI_Comm &       mpi_comm = PETSC_COMM_WORLD)
+    {
+      AssertTS(TSCreate(mpi_comm, &ts));
+      AssertTS(TSSetApplicationContext(ts, this));
+      reinit(data);
+    }
+
+    /**
+     * Destructor.
+     */
+    virtual ~TimeStepper()
+    {
+      AssertTS(TSDestroy(&ts));
+    }
+
+    /**
+     * Return the PETSc TS object.
+     */
+    TS
+    petsc_ts()
+    {
+      return ts;
+    }
+
+    /**
+     * Return a reference to the MPI communicator object in use with this
+     * object.
+     */
+    const MPI_Comm &
+    get_mpi_communicator() const
+    {
+      static MPI_Comm comm = PETSC_COMM_SELF;
+      MPI_Comm pcomm       = PetscObjectComm(reinterpret_cast<PetscObject>(ts));
+      if (pcomm != MPI_COMM_NULL)
+        comm = pcomm;
+      return comm;
+    }
+
+    /**
+     * Reset the subobjects associated with TS, do not change customization.
+     */
+    void
+    reinit()
+    {
+      AssertTS(TSReset(ts));
+    }
+
+    /**
+     * Reset the subobjects associated with TS, changes customization according
+     * to data.
+     */
+    void
+    reinit(const TimeStepperData &data)
+    {
+      reinit();
+
+      // Solver type
+      if (data.tstype.size())
+        AssertTS(TSSetType(ts, data.tstype.c_str()));
+
+      // Time and steps limits
+      AssertTS(TSSetTime(ts, data.initial_time));
+      if (data.final_time > data.initial_time)
+        AssertTS(TSSetMaxTime(ts, data.final_time));
+      if (data.initial_step_size > 0.0)
+        AssertTS(TSSetTimeStep(ts, data.initial_step_size));
+      if (data.max_steps >= 0)
+        AssertTS(TSSetMaxSteps(ts, data.max_steps));
+      // Decide how to end the integration. Either stepover the final time or
+      // match it.
+      AssertTS(TSSetExactFinalTime(ts,
+                                   data.match_step ?
+                                     TS_EXACTFINALTIME_MATCHSTEP :
+                                     TS_EXACTFINALTIME_STEPOVER));
+
+      // Adaptive tolerances
+      PetscReal atol = data.absolute_tolerance > 0.0 ?
+                         data.absolute_tolerance :
+                         static_cast<PetscReal>(PETSC_DEFAULT);
+      PetscReal rtol = data.relative_tolerance > 0.0 ?
+                         data.relative_tolerance :
+                         static_cast<PetscReal>(PETSC_DEFAULT);
+      AssertTS(TSSetTolerances(ts, atol, nullptr, rtol, nullptr));
+
+      // Adaptive time stepping
+      TSAdapt tsadapt;
+      AssertTS(TSGetAdapt(ts, &tsadapt));
+
+      // Use adaptive time stepping only if explicitly requested
+      if (data.absolute_tolerance == 0.0 && data.relative_tolerance == 0.0)
```

Comment:
This is an odd choice of condition to indicate that users want to use adaptive time stepping. Would it make sense to introduce an explicitly boolean flag to say so? Alternatively, the condition should at least be documented.